src/core/llm/services/vercel.ts

```diff
@@ -92,14 +95,53 @@ export class VercelLLMService implements ILLMService {
                     inputSchema: jsonSchema(tool.parameters),
-                    execute: async (args: unknown) => {
+                    execute: async (args: unknown, _options?: { toolCallId?: string }) => {
                         try {
-                            return await this.toolManager.executeTool(
+                            const rawResult = await this.toolManager.executeTool(
                                 toolName,
                                 args as Record<string, unknown>,
                                 this.sessionId
                             );
+
+                            // Queue raw, unfiltered result under the specific toolCallId
+                            const callId = _options?.toolCallId ?? this.sessionId;
+                            const queue = this.rawResultsByCallId.get(callId) ?? [];
+                            queue.push(rawResult);
+                            this.rawResultsByCallId.set(callId, queue);
+
+                            // Sanitize tool result to prevent large/base64 media from exploding context
+                            // Convert arbitrary result -> InternalMessage content (media as structured parts)
+                            // then summarize to concise text suitable for Vercel tool output.
+                            try {
+                                const safeContent = sanitizeToolResultToContent(rawResult);
+                                const summaryText = summarizeToolContentForText(safeContent);
+                                return summaryText;
+                            } catch (sanErr) {
+                                // Fallback: stringify/truncate to keep context safe
+                                const asString =
+                                    typeof rawResult === 'string'
+                                        ? rawResult
+                                        : (() => {
+                                              try {
+                                                  return JSON.stringify(rawResult);
+                                              } catch {
+                                                  return String(rawResult);
+                                              }
+                                          })();
+                                const MAX_LEN = 4000;
+                                const truncated =
+                                    asString.length > MAX_LEN
+                                        ? `${asString.slice(0, MAX_LEN)}... [truncated]`
+                                        : asString;
+                                logger.warn(
+                                    `Vercel tool result sanitization failed for ${toolName}: ${String(
+                                        sanErr
+                                    )}. Returning truncated text.`
+                                );
+                                return truncated;
+                            }
```

Comment:
we already handle all of this error stuff in the helper functions so we don't need this catch

`sanitizeToolResultToContent` is already a safe function that catches errors, and handles them safely like this, so it never errors out so we don't need this

we can make the summarize func also safe in simialr manner